src/io/dataset.cc

```diff
@@ -0,0 +1,696 @@
+    std::string img_list;
+    /*! \brief the path separator character, by default it's ; */
+    char path_sep;
+    /*! \brief If flag is 0, always convert to grayscale(1 channel).
+    * If flag is 1, always convert to colored (3 channels).
+    * If flag is -1, keep channels unchanged.
+    */
+    int flag;
+    // declare parameters
+    DMLC_DECLARE_PARAMETER(ImageSequenceDatasetParam) {
+        DMLC_DECLARE_FIELD(img_list)
+            .describe("The list of image absolute paths.");
+        DMLC_DECLARE_FIELD(path_sep).set_default('|')
+            .describe("The path separator for joined image paths.");
+        DMLC_DECLARE_FIELD(flag).set_default(1)
+            .describe("If 1, always convert to colored, if 0 always convert to grayscale.");
+    }
+};  // struct ImageSequenceDatasetParam
+
+DMLC_REGISTER_PARAMETER(ImageSequenceDatasetParam);
+
+class ImageSequenceDataset : public Dataset {
+  public:
+    ImageSequenceDataset* Clone(void) const {
+      return new ImageSequenceDataset(*this);
+    }
+
+    void Init(const std::vector<std::pair<std::string, std::string> >& kwargs) {
+      std::vector<std::pair<std::string, std::string> > kwargs_left;
+      param_.InitAllowUnknown(kwargs);
+      img_list_ = dmlc::Split(param_.img_list, param_.path_sep);
+    }
+
+    uint64_t GetLen() const {
+      return img_list_.size();
+    }
+
+    bool GetItem(uint64_t idx, std::vector<NDArray>& ret) {
+#if MXNET_USE_OPENCV
+      CHECK_LT(idx, img_list_.size())
+        << "GetItem index: " << idx << " out of bound: " << img_list_.size();
+      cv::Mat res = cv::imread(img_list_[idx], param_.flag);
+      CHECK(!res.empty()) << "Decoding failed. Invalid image file.";
+      const int n_channels = res.channels();
+      ret.resize(1);
+      if (n_channels == 1) {
+        SwapImageChannels<1>(res, ret[0]);
+      } else if (n_channels == 3) {
+        SwapImageChannels<3>(res, ret[0]);
+      } else if (n_channels == 4) {
+        SwapImageChannels<4>(res, ret[0]);
+      }
+      return true;
+#else
+    LOG(FATAL) << "Opencv is needed for image decoding.";
+#endif
+    return false;
+    };
+
+  private:
+    /*! \brief parameters */
+    ImageSequenceDatasetParam param_;
+    /*! \brief image list */
+    std::vector<std::string> img_list_;
+};
+
+MXNET_REGISTER_IO_DATASET(ImageSequenceDataset)
+ .describe("Image Sequence Dataset")
+ .add_arguments(ImageSequenceDatasetParam::__FIELDS__())
+ .set_body([]() {
+     return new ImageSequenceDataset();
+});
+
+struct NDArrayDatasetParam : public dmlc::Parameter<NDArrayDatasetParam> {
+    /*! \brief the source ndarray */
+    std::intptr_t arr;
+    // declare parameters
+    DMLC_DECLARE_PARAMETER(NDArrayDatasetParam) {
+        DMLC_DECLARE_FIELD(arr)
+            .describe("Pointer to NDArray.");
+    }
+};  // struct NDArrayDatasetParam
+
+DMLC_REGISTER_PARAMETER(NDArrayDatasetParam);
+
+class NDArrayDataset : public Dataset {
```

Comment:
Specify all datasets as `final`? This will allow the compiler to perform further optimization